### PR TITLE
chore: Configure DWI with tlsCertificateConfigmapRef when certificate…

### PR DESCRIPTION
**Pull request by tolusha**
…s imported to che-operator

### What does this PR do?
This PR configures the DevWorkspace Operator to use Che's merged CA bundle for TLS by setting tlsCertificateConfigmapRef in the DWI routing config when custom certificates are imported, and adding a watch label so DWO  picks up changes. It also refactors the TLS certificate reconciliation to use newer client APIs, adds owner references, removes deprecated sync helpers, and splits a large test file into domain-specific ones.

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

https://github.com/eclipse-che/che/issues/23870

### How to test this PR?

1. Deploy the operator:
`chectl server:deploy -p openshift --che-operator-image quay.io/abazko/che-operator:next`

2. Create k8s resources to host a parent devfile:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: parent
---
apiVersion: v1
kind: Pod
metadata:
  name: server
  namespace: parent
  labels:
    app: file-server-app
spec:
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
  containers:
    - name: httpd
      image: 'quay.io/abazko/operator:parent-devfile'
      ports:
        - containerPort: 8000
      securityContext:
        allowPrivilegeEscalation: false
        capabilities:
          drop:
            - ALL
---
apiVersion: v1
kind: Service
metadata:
  name: file-server
  namespace: parent
spec:
  selector:
    app: file-server-app
  ports:
    - protocol: TCP
      port: 80
      targetPort: 8000
---
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  name: parent
  namespace: parent
spec:
  to:
    name: file-server
    kind: Service
  tls:
    insecureEdgeTerminationPolicy: Redirect
    termination: edge
  port:
    targetPort: 8000
```

3. Ensure parent devfile is accessible by url
`https://parent-parent.<openshift-base-domain>/devfile.yaml`  (for instance `https://parent-parent.apps-crc.testing/devfile.yaml`)

4. Import certificate into Che
```
openssl s_client -connect parent-parent.apps-crc.testing:443 -showcerts < /dev/null | sed -n '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/p' > parent.crt

kubectl create configmap custom-ca-certificates \
    --from-file=parent.crt \
    --namespace=eclipse-che

kubectl label configmap custom-ca-certificates \
    app.kubernetes.io/component=ca-bundle \
    app.kubernetes.io/part-of=che.eclipse.org  \
    --namespace=eclipse-che
```

5. Check that DWO is configured 
`oc get dwoc -n eclipse-che devworkspace-config -o jsonpath="{.config.routing}"`

6. Start a workspace from `https://gist.githubusercontent.com/tolusha/6c68d907a52d235b311f196478ce08b6/raw/d94a38042093d9a5fe021091d861cb9ea15ac78e/devfile.yaml`

#### Common Test Scenarios
- [x] Deploy Eclipse Che
- [x] Start an empty workspace
- [ ] Open terminal and build/run an image
- [ ] Stop a workspace
- [x] Check operator logs for reconciliation errors or infinite reconciliation loops

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by tolusha — COMMENTED** [review]
Comprehensive review completed (standard + deep + impact analysis). The PR correctly implements the feature and addresses issue #23870. 

**Integration contracts verified** against DevWorkspace Operator source code:
- ✅ `TLSCertificateConfigmapRef` field structure matches DWO expectations
- ✅ `controller.devfile.io/watch-configmap` label verified in DWO constants

**Positive observations:**
- Clean abstraction improvement with `ConfigMapEnsureMetadata(obj)`
- Good test coverage with both create and update scenarios  
- Correct ordering prevents nil pointer dereference
- Operationally sound - proper resource cleanup, error handling
- No security concerns - improves security posture

**Minor observations** are posted as inline comments for discussion.

**Review by tolusha — COMMENTED** [review]
## Review Summary

Comprehensive review complete across standard, deep, and impact analysis. The PR correctly implements the TLSCertificateConfigmapRef feature and solves the linked issue. Integration contracts with DevWorkspace Operator are verified. 

**Positive highlights:**
- Thorough test coverage with 4 well-structured test cases
- Clean `GetLabelsAndAnnotations` abstraction eliminates hardcoded label lists
- Consistent application of the new pattern across all call sites
- Integration contracts correctly implemented (verified against DWO source)

**Key recommendations:**
1. Add cleanup path for stale TLSCertificateConfigmapRef when ConfigMap is removed
2. Make test assertions more precise instead of blanket-ignoring Routing field
3. Remove dead code `ConfigMapEnsureLabels`
4. Document behavioral change in diff comparison logic for release notes

See inline comments for details. None are blocking - the core feature is well-implemented.

---
📋 **Review artifacts:**
- Standard review: ✅ Complete
- Deep review: ✅ Complete  
- Impact review: ✅ Complete
- Verdict: 💬 **Comment** - Ready to merge with recommendations addressed

**Review by tolusha — COMMENTED** [review]
## PR Review Summary

Three comprehensive reviews completed (standard, deep, and impact analysis). Overall verdict: **Well-structured PR with good test coverage**. The code properly addresses issue #23870 by auto-configuring DWO's TLS certificate reference.

### Key Findings

**Strengths:**
- Comprehensive test coverage (7 lifecycle test cases)
- Clean `GetLabelsAndAnnotations` + `ConfigMapWithMetadata` refactoring
- Proper bidirectional sync (set when exists, clear when absent)
- Good separation of concerns

**Suggestions for improvement:**
1. Test coverage masks `Routing` field - consider explicit assertions
2. `ConfigMapEnsureLabels` may be dead code - verify and remove if unused
3. Add combined proxy + TLS integration test
4. Verify owner reference addition works on upgrade path
5. Add idempotency test case
6. Document `GetLabelsAndAnnotations` invariant

**Impact Review:**
- Owner reference addition improves garbage collection ✅
- No security, observability, or resource cleanup concerns
- One item to verify: owner reference addition on existing ConfigMaps during upgrade

See inline comments for details on each finding.

---
🤖 Automated review by che-ai-assistant

**Review by tolusha — COMMENTED** [review]
## Review Summary

This PR successfully addresses eclipse-che/che#23870 by configuring DWO to reference the merged CA bundle ConfigMap for TLS certificate trust. The implementation is solid with excellent test coverage (~500 new test lines covering 12 scenarios).

**Verdict**: ✅ Approve - No blocking issues. The warnings below are minor defensive coding considerations.

### Key Points

1. **Sync method change** (pkg/deploy/tls/certificates.go:823-837) - Verify ClientWrapper.Sync provides same semantics as deploy.Sync
2. **OpenShift CA Data ignore** (pkg/deploy/tls/certificates.go:179-188) - Already tested, but verify Sync uses server-side apply vs full replace
3. **Zero-check fragility** (pkg/deploy/devworkspace/dev_workspace_config.go:332-335) - Struct comparison fragile if new pointer/slice fields added
4. **Missing comment** (pkg/deploy/tls/certificates.go:792) - Add context for watch-configmap label
5. **Cross-package docs** - Document TLS → DWO data flow connection

### Positive Feedback

- Excellent test coverage (12 scenarios including edge cases)
- Clean refactoring from DefaultsLabelKeys to GetLabelsAndAnnotations
- Proper cleanup logic preventing empty routing config
- Data-ignore fix prevents infinite reconciliation loop - operational improvement!

---

Full review findings saved to: `/home/user/.claude/ok-pr-review/repos/eclipse-che/che-operator/2137-{review,deep-review,impact}.md`

**Review by tolusha — COMMENTED** [review]
## Code Review Summary

Reviewed all changes across standard, deep, and impact analysis. The core TLS integration feature is well-implemented with excellent test coverage. Here are the findings:

### Issues to Address

**1. Stale test assertion (pkg/deploy/server/chehost_reconciler_test.go:57)**
The `deploy.Get()` call was replaced with `ctx.ClusterAPI.Client.Get()` which returns only `error`, not `(bool, error)`. Line 57's `assert.True(t, done)` now asserts a stale value from line 52 rather than testing the service was found. The `done` variable should be removed since `assert.Nil(t, err)` already catches failures.

**2. Missing owner reference (pkg/deploy/tls/certificates.go:307)**
`syncSelfSignedCertificates` creates a ConfigMap but doesn't call `controllerutil.SetControllerReference()`, unlike other sync functions modified in this PR. Without this, the ConfigMap won't be garbage-collected when CheCluster is deleted. Consider adding the owner reference for consistency.

**3. Pre-existing test bug (pkg/deploy/devworkspace/dev_workspace_config_scheduling_test.go:1110)**
`TestReconcileDevWorkspaceConfigRuntimeClassName` asserts `SchedulerName` instead of `RuntimeClassName`. This was in the original test file (line 4710) and means RuntimeClassName has zero effective test coverage. Good opportunity to fix during refactoring:
```go
assert.Equal(t, testCase.expectedOperatorConfig.Workspace.RuntimeClassName, dwoc.Config.Workspace.RuntimeClassName)
```

**4. Missing test for watch-configmap label**
The `controller.devfile.io/watch-configmap=true` label (certificates.go:416) is critical for DWO to watch certificate changes, but has no dedicated test. Consider adding verification in certificates_test.go.

**5. Test blind spot from IgnoreFields**
Several tests (storage_test.go:3111, scheduling_test.go:811, security_test.go:1521) ignore `TLSCertificateConfigmapRef` via `cmpopts.IgnoreFields`. While necessary, this creates a blind spot - tests won't catch unintended TLS ref mutations. Consider explicit nil assertions where no CA bundle is provided.

### Suggestions

**reflect.DeepEqual fragility (dev_workspace_config.go:343)**
Using `reflect.DeepEqual(operatorConfig.Routing, &controllerv1alpha1.RoutingConfig{})` couples cleanup to zero-values of all fields. More explicit:
```go
if operatorConfig.Routing.TLSCertificateConfigmapRef == nil && 
   operatorConfig.Routing.ProxyConfig == nil && 
   operatorConfig.Routing.DefaultRoutingClass == "" {
    operatorConfig.Routing = nil
}
```

### Excellent Work

- **Thorough TLS test coverage**: 8 test cases + 4 proxy composition tests covering all scenarios
- **Clean test reorganization**: Splitting 3,328-line monolith into 6 topic-focused files improves maintainability significantly  
- **Good API cleanup**: Replacing `deploy.Get()`, `GetForClient()`, `DeleteClusterObject()` and `DefaultsLabelKeys` with `ClientWrapper` methods and dynamic `GetLabelsAndAnnotations()`
- **DWO watch label**: Adding `controller.devfile.io/watch-configmap=true` ensures certificate propagation without restarts
- **Proper error wrapping**: Consistent `fmt.Errorf("...: %w", err)` throughout

### System-Level Assessment

✅ **No operational concerns** - Feature activates automatically on upgrade, is idempotent, and introduces no new failure modes. Reconciliation stability verified by tests. No backward compatibility issues (removed exports are internal-only).

Great implementation overall! The issues above are minor - #1 and #3 are test fixes, #2 is for consistency, #4-5 are nice-to-haves for test robustness.

---
🤖 Automated review by [ok-pr-review](https://github.com/ok-plugins/ok-pr-review)

**Review by tolusha — COMMENTED** [review]
## Comprehensive Review: eclipse-che/che-operator#2137

✅ **LGTM with minor suggestions** - This PR correctly implements TLS certificate ConfigMap reference for DevWorkspace Operator integration. All three review levels (standard, deep, impact) found no blocking issues.

### Summary

This PR enables the che-operator to automatically configure DevWorkspace Operator with trusted CA certificates by setting `tlsCertificateConfigmapRef` when the merged CA bundle ConfigMap exists. Additionally, it refactors the DevWorkspace config tests by splitting a 3328-line monolithic test file into 6 focused files organized by functional area.

**Linked issue:** https://github.com/eclipse-che/che/issues/23870

### Key Strengths

- **Excellent test coverage**: 12 test cases (8 TLS + 4 composition) covering all state transitions
- **Proper resource lifecycle**: `SetControllerReference` added to operator-created ConfigMaps for garbage collection
- **Consistent refactoring**: Uniform migration from old sync utilities to new client wrapper pattern
- **Clean test organization**: Monolithic 3328-line test file split into 6 focused files
- **Defensive nil-guarding**: Three-layer defense prevents nil panics across all reconciliation states
- **Integration correctness**: All DWO contracts verified against actual types
- **Operational safety**: No performance concerns, no security risks, fully automatic upgrade path

### Review Findings

I've added inline comments on specific areas for consideration. All are suggestions or minor concerns - none are blocking.

---

📊 **Review Statistics**
- Standard review: ✅ Approve
- Deep review: ✅ Design is sound
- Impact review: ✅ No concerns
- Files reviewed: 16/16
- Test cases added: 12
- Owner references improved: 4 ConfigMaps

🤖 Generated by Claude Code PR Review

**Comment by tolusha** [comment]
### 🔍 Review: Verify mandatory label keys for selfSignedCertCM

**Location:** `pkg/deploy/tls/certificates.go` around line 754

The `mandatoryLabelKeys` for `selfSignedCertCM` includes `ConfigOpenShiftIOInjectTrustedCaBundle`, but the ConfigMap blueprint doesn't actually set this label:

```go
mandatoryLabelKeys := slices.Collect(maps.Keys(defaultLabels))
mandatoryLabelKeys = append(mandatoryLabelKeys, constants.ConfigOpenShiftIOInjectTrustedCaBundle)
```

If this label exists on the actual ConfigMap from a previous reconcile but not in the blueprint, the diff comparator might trigger unnecessary updates. Double-check that `cmpMetadata` handles this correctly - if it only compares keys present in both objects, this should be fine. Otherwise, consider only including labels that are actually set on the blueprint.

**Review by tolusha — COMMENTED** [review]
## Comprehensive PR Review: Configure DWI with tlsCertificateConfigmapRef

I've completed a comprehensive review including standard code review, deep design analysis, and system-level impact assessment. Overall, this is a well-implemented PR with thorough test coverage that correctly addresses the linked issue.

### ✅ Summary

The PR successfully configures DevWorkspace Operator to trust CA certificates imported by che-operator by:
- Adding `updateTLSCertificateConfigmapRef` to set `TLSCertificateConfigmapRef` when `ca-certs-merged` exists with data
- Adding the `controller.devfile.io/watch-configmap=true` label so DWO can watch certificate changes
- Adding `SetControllerReference` to 4 ConfigMaps for proper garbage collection

### 📋 Findings

#### ⚠️ Items to Verify Before Merge

1. **Owner reference on OpenShift CA bundle ConfigMap** (`certificates.go:177`)
   - Adding `SetControllerReference` means the `ca-bundle` ConfigMap will be garbage-collected when the CheCluster CR is deleted
   - On OpenShift, this ConfigMap may have existed before CheCluster and might be expected to survive deletion
   - Please verify this is the desired behavior

2. **No watch on `ca-certs-merged` for che-operator** (`dev_workspace_config.go:103`)
   - Certificate changes won't propagate to DWO config until the next unrelated reconciliation trigger
   - Consider adding a watch/informer for timely propagation or verify this delay is acceptable

#### 💡 Suggestions (Non-blocking)

3. **Test masking may hide cross-feature interference** (`dev_workspace_config_scheduling_test.go:803`)
   - Four test functions use `cmpopts.IgnoreFields(controllerv1alpha1.RoutingConfig{}, "TLSCertificateConfigmapRef")`
   - While routing tests cover TLS thoroughly, this means tests won't detect if TLS accidentally interferes with other features
   - Consider `assert.Equal` on specific fields instead of broad ignore patterns

4. **Check for ConfigMapEnsureLabels dead code** (`diffs.go:44`)
   - `DefaultsLabelKeys` was removed but `ConfigMapEnsureLabels` may still exist
   - Verify with: `grep -r "ConfigMapEnsureLabels" --include="*.go"`

5. **Add error-path test** (`dev_workspace_config.go:103`)
   - Test case where `GetIgnoreNotFound` returns an error would improve coverage

6. **Document TypeMeta removal rationale**
   - TypeMeta setting was removed in multiple places (certificates.go lines 131-134, 240-243, 337-340)
   - A comment explaining that `ClientWrapper.Sync` handles this would help future maintainers

7. **Clarify test expectations** (`dev_workspace_config_routing_test.go:347`)
   - Inconsistency between `nil` and `&RoutingConfig{}` expectations in "clear" tests
   - Comment explaining the implicit behavior would be helpful

### 🎉 Excellent Work

- **Comprehensive test coverage**: 8 TLS test cases + 4 proxy composition tests
- **Clean refactoring**: 3328-line test file split into 6 focused files
- **Thoughtful design**: `GetLabelsAndAnnotations` is a genuine improvement
- **Operational improvement**: Proper garbage collection via controller references
- **Integration detail**: `DevWorkspaceWatchConfigMapLabel` enables DWO to react to updates

### Verdict

💬 **Comment** - Ready to merge after verifying items 1 and 2 above. The suggestions are non-blocking improvements.

**Review by tolusha — COMMENTED** [review]
## Review Summary

Great work on this PR! The implementation correctly solves the linked issue with comprehensive test coverage. I've left a few suggestions for defensive coding and documentation improvements, plus some positive notes on what was done well.

### Key Findings:
- ✅ Feature implementation is correct and complete
- ✅ Test coverage is excellent (8 scenarios for TLS cert ref)
- ✅ Owner reference additions improve resource lifecycle
- 💡 A few defensive coding suggestions
- 💡 Minor documentation improvements for maintainability

All feedback is non-blocking - the PR is ready to merge as-is.

File: pkg/deploy/labels.go [hunk 1]
```diff
@@ -61,6 +65,10 @@ func GetLegacyLabels(component string) map[string]string {
 	}
 }
 
+func GetLabelsAndAnnotations(obj client.Object) ([]string, []string) {
```
Comment (tolusha) [inline pkg/deploy/labels.go:70]: **Consider documenting non-deterministic key order**

The `GetLabelsAndAnnotations` function uses `slices.Collect(maps.Keys(...))` which returns keys in map iteration order (non-deterministic). While this is fine for the current use case, a brief doc comment noting the order is non-deterministic would prevent future misuse.

```suggestion
// GetLabelsAndAnnotations extracts label and annotation keys from an object.
// Note: key order is non-deterministic (map iteration order).
func GetLabelsAndAnnotations(obj client.Object) ([]string, []string) {
```

File: pkg/deploy/tls/certificates.go [hunk 2]
```diff
@@ -156,32 +150,35 @@ func (c *CertificatesReconciler) syncOpenShiftCABundleCertificates(ctx *chetypes
...
-		// Ignore Data field to allow OpenShift network operator inject certificates into CM
-		// and avoid endless reconciliation loop
-		return deploy.Sync(
-			ctx,
-			openShiftCaBundleCM,
-			diffs.ConfigMap(append(deploy.DefaultsLabelKeys, constants.ConfigOpenShiftIOInjectTrustedCaBundle), nil),
-		)
+	if err := controllerutil.SetControllerReference(ctx.CheCluster, openShiftCaBundleCM, ctx.ClusterAPI.Scheme); err != nil {
+		return false, err
 	}
```
Comment (tolusha) [inline pkg/deploy/tls/certificates.go:168]: **Defensive copy for slice append safety**

The `append` to `deploy.GetLabelKeys()` is currently safe because `GetLabelKeys()` returns a new slice each call. However, if `GetLabelKeys()` is ever refactored to return a cached slice, this would silently corrupt it. Consider using `slices.Concat` for defensive safety:

```suggestion
mandatoryLabelKeys := slices.Concat(deploy.GetLabelKeys(), []string{constants.ConfigOpenShiftIOInjectTrustedCaBundle})
```

**Comment by tolusha** [comment]
Hi! I'm **che-ai-assistant** — I help with your pull requests.

**Available commands**:
- `/che-ai-assistant generate-che-doc` — Generate a documentation PR based on this PR's changes
- `/che-ai-assistant ok-pr-review` — Run a comprehensive PR review (summary, code review, deep review, impact analysis)
- `/che-ai-assistant help` — Show this help message

**Comment by tolusha** [comment]
/che-ai-assistant ok-pr-review 

Review is complete. Please check the review comments below.

**Review by tolusha — COMMENTED** [review]
## Comprehensive Review Summary

I've completed a thorough multi-dimensional review of this PR:
- ✅ **Standard Review**: Approve - ready to merge
- ✅ **Deep Review**: Design is sound
- ✅ **Impact Review**: No system-level concerns

## Key Strengths

**Excellent test coverage**: 8 test cases for the new TLS ConfigMap ref feature covering all edge cases (creation, update, removal, empty data, idempotency). The composition tests for proxy + TLS settings show attention to integration concerns.

**Clean test file organization**: Breaking the 3328-line monolithic test into 6 domain-specific files (routing, scheduling, security, service account, storage, workspace) greatly improves maintainability.

**Proper resource lifecycle**: Owner references ensure ConfigMaps are garbage-collected when CheCluster is deleted. The deliberate exclusion of admin-created resources (git trusted certs) shows understanding of K8s ownership semantics.

**Type-safe DWO integration**: The `controller.devfile.io/watch-configmap=true` label and `TLSCertificateConfigmapRef` use imported constants from the DevWorkspace Operator module - no magic strings.

**Consistent refactoring**: Migration from deprecated `deploy.Get()` to `ClientWrapper.GetIgnoreNotFound()` is applied uniformly across all call sites.

## Minor Suggestions

1. **Test coverage gap**: The `updateTLSCertificateConfigmapRef` error path (when `GetIgnoreNotFound` fails) isn't exercised in tests. Consider adding a test case that simulates a client error when fetching the ConfigMap.

2. **Diff pattern documentation**: In `syncCheCABundleCerts` (certificates.go:455), the shift to using `GetLabelsAndAnnotations(mergedCABundlesCM)` for diff keys is clever but creates an implicit dependency on label ordering. Consider adding a comment documenting that labels must be set on the blueprint before this call.

3. **Optional helper extraction**: The pattern of `SetControllerReference` + `ClientWrapper.Sync` + `return err == nil, err` appears in 5 methods. A `SyncOwned(ctx, owner, blueprint, opts)` helper could reduce boilerplate, though the current code is correct.

## Verdict

**✅ Approve** - This PR is well-designed, thoroughly tested, and ready to merge. The TLS certificate integration is correctly implemented and the refactoring away from deprecated helpers is clean and consistent.

---

📊 **Review completed**: Standard + Deep + Impact analysis

**Comment by tolusha** [comment]
/che-ai-assistant generate-che-doc

Created documentation PR: https://github.com/eclipse-che/che-docs/pull/3128

File: pkg/deploy/devworkspace/dev_workspace_config_security_test.go [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+				fmt.Sprintf("Did not get expected ContainerSecurityContext.\nDiff:%s", cmp.Diff(expectedContainerSecurityContext, actualContainerSecurityContext)))
+		})
+	}
+}
+
+func TestReconcileDevWorkspacePodSecurityContext(t *testing.T) {
+	type testCase struct {
+		name                   string
+		cheCluster             *chev2.CheCluster
+		existedObjects         []client.Object
```
Comment (rohanKanojia) [inline pkg/deploy/devworkspace/dev_workspace_config_security_test.go:577]: I notice in `TestReconcileDevWorkspaceContainerSecurityContext` and `TestReconcileDevWorkspacePodSecurityContext` test cases define existedObjects, but the test context is created with 
```go
deployContext := test.NewCtxBuilder().WithCheCluster(testCase.cheCluster).Build()
```
Should deployContext include existedObjects so that updated paths are also tested?

File: pkg/deploy/devworkspace/dev_workspace_config_security_test.go [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+func TestReconcileDevWorkspaceConfigForContainerCapabilities(t *testing.T) {
+	type testCase struct {
+		name                   string
+		cheCluster             *chev2.CheCluster
+		existedObjects         []client.Object
+		expectedOperatorConfig *controllerv1alpha1.OperatorConfiguration
+	}
+
+	var unmasked = corev1.UnmaskedProcMount
+	var testCases = []testCase{
```
Comment (rohanKanojia) [inline pkg/deploy/devworkspace/dev_workspace_config_security_test.go]: Nit (optional): `testCases := []testCase{}` could be used here since the type is already apparent from the composite literal, though this is mostly a style preference.

File: pkg/deploy/devworkspace/dev_workspace_config_service_account_test.go [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+func TestReconcileServiceAccountConfig(t *testing.T) {
+	type testCase struct {
+		name                   string
+		cheCluster             *chev2.CheCluster
+		expectedOperatorConfig *controllerv1alpha1.OperatorConfiguration
+	}
+
+	var testCases = []testCase{
+		{
+			name: "Case #1",
```
Comment (rohanKanojia) [inline pkg/deploy/devworkspace/dev_workspace_config_service_account_test.go]: Nit: Could we use descriptive test names instead of Case #1, Case #2, etc.? It makes failures much easier to interpret from the test output without opening the test file.

```suggestion
            name: "ServiceAccount configured with auto-provision enabled",
```

**Comment by tolusha** [comment]
@rohanKanojia 
fixed

**Comment by tolusha** [comment]
closed in favor of https://github.com/eclipse-che/che-operator/pull/2146
